File: src/csharp/SlimShader/Chunks/Ifce/InterfaceSlot.cs
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SlimShader.Util;

namespace SlimShader.Chunks.Ifce
{
	public class InterfaceSlot
	{
		public uint StartSlot { get; set; }
		public uint SlotSpan { get; private set; }
		public List<uint> TypeIDs { get; private set; }
		public List<uint> TableIDs { get; private set; }

		public InterfaceSlot()
		{
			TypeIDs = new List<uint>();
			TableIDs = new List<uint>();
		}

		public static InterfaceSlot Parse(BytecodeReader reader, BytecodeReader interfaceSlotReader)
		{
			var slotSpan = interfaceSlotReader.ReadUInt32();

			var count = interfaceSlotReader.ReadUInt32();

			var typeIDsOffset = interfaceSlotReader.ReadUInt32();
			var typeIDsReader = reader.CopyAtOffset((int) typeIDsOffset);

			var tableIDsOffset = interfaceSlotReader.ReadUInt32();
			var tableIDsReader = reader.CopyAtOffset((int) tableIDsOffset);

			var result = new InterfaceSlot
			{
				SlotSpan = slotSpan
			};

			for (int i = 0; i < count; i++)
			{
				result.TypeIDs.Add(typeIDsReader.ReadUInt16());
				result.TableIDs.Add(tableIDsReader.ReadUInt32());
			}

			return result;
		}

		public override string ToString()
		{
			// For example:
			// | Type ID  |   0     |0    1    2    
			// | Table ID |         |0    1    2    
			// +----------+---------+---------------------------------------

			var sb = new StringBuilder();

			string slotSpan = (SlotSpan == 1) 
				? StartSlot.ToString() : 
				string.Format("{0}-{1}", StartSlot, StartSlot + SlotSpan - 1);
			sb.AppendLine(string.Format("// | Type ID  |   {0,-5} |{1}", slotSpan, 
				string.Join(" ", TypeIDs.Select(x => string.Format("{0,-4}", x)))));
			sb.AppendLine(string.Format("// | Table ID |         |{0}",
				string.Join(" ", TableIDs.Select(x => string.Format("{0,-4}", x)))));
			sb.AppendLine("// +----------+---------+---------------------------------------");
			return sb.ToString();
		}
	}
}

File: src/cpp/Source/Chunks/Ifce/InterfaceSlot.cpp
#include "PCH.h"
#include "InterfaceSlot.h"

#include <sstream>

using namespace std;
using namespace SlimShader;

InterfaceSlot InterfaceSlot::Parse(const BytecodeReader& reader, BytecodeReader& interfaceSlotReader)
{
	assert(interfaceSlotReader.ReadUInt32() == 1); // Unknown

	auto count = interfaceSlotReader.ReadUInt32();

	auto typeIDsOffset = interfaceSlotReader.ReadUInt32();
	auto typeIDsReader = reader.CopyAtOffset(typeIDsOffset);

	auto tableIDsOffset = interfaceSlotReader.ReadUInt32();
	auto tableIDsReader = reader.CopyAtOffset(tableIDsOffset);

	InterfaceSlot result;

	vector<uint32_t> typeIDs, tableIDs;
	for (uint32_t i = 0; i < count; i++)
	{
		result._typeIDs.push_back(typeIDsReader.ReadUInt16());
		result._tableIDs.push_back(tableIDsReader.ReadUInt32());
	}

	return result;
}

const uint32_t InterfaceSlot::GetID() const { return _id; }
void InterfaceSlot::SetID(uint32_t id) { _id = id; }

const std::vector<uint32_t>& InterfaceSlot::GetTypeIDs() const { return _typeIDs; }
const std::vector<uint32_t>& InterfaceSlot::GetTableIDs() const { return _tableIDs; }

std::ostream& SlimShader::operator<<(std::ostream& out, const InterfaceSlot& value)
{
	// For example:
	// | Type ID  |   0     |0    1    2    
	// | Table ID |         |0    1    2    
	// +----------+---------+---------------------------------------

	stringstream typeIDsStream;
	for (auto typeID : value._typeIDs)
		typeIDsStream << boost::format("%-4i ") % typeID;

	stringstream tableIDsStream;
	for (auto tableID : value._tableIDs)
		tableIDsStream << boost::format("%-4i ") % tableID;

	out << (boost::format("// | Type ID  |   %i     |%s")
		% value._id
		% typeIDsStream.str())
		<< endl;
	out << (boost::format("// | Table ID |         |%s")
		% tableIDsStream.str())
		<< endl;
	out << "// +----------+---------+---------------------------------------" << endl;

	return out;
}